[Engulfy] iP

Engulfy opened the pull request:
# 🌊 Engulfy Bot

> *"Engulf your tasks before they engulf you! 🔥"*  - Engulfy 

## Features
- ✅ **Add Tasks: Create todos, deadlines, and events with ease.** 📜
- ✅ **Mark & Unmark: Stay on top of your progress by marking tasks as done or undone.** ✅
- ✅ **Delete Tasks: Remove tasks that are no longer needed.** ❌
- ✅ **Find Tasks: Search for tasks containing specific keywords.** 🔍
- ✅ **Persistent Storage: Never lose your tasks! Engulfy saves them automatically.** 💾

## Getting Started
1. Download it from [this link](https://github.com/Engulfy/ip) .
2. Run it by double-clicking the file.
3. Start adding your tasks!
4. To-Do List ✅
- [ ] Implement basic task management  
- [ ] Add unit tests  
- [ ] Implement GUI  


### Code Example
```java
public class Main {
    public static void main(String[] args) {
           new engulfy.command.Engulfy().run();
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Engulfy committed:
Renamed bot, updated greet and exit

Engulfy committed:
Echo function added

Engulfy committed:
Add list functionality added

Engulfy committed:
Added Mark as Done Functionality

Engulfy committed:
Added Todo, Event, Deadline

Engulfy committed:
Added Automated Text UI Testing

Engulfy committed:
Added Error handling capabilities

Engulfy committed:
Added delete functionalities

Engulfy committed:
Implemented file saving

Engulfy committed:
Merge Level 7 implementation

Engulfy committed:
Implemented Level 8 which can identify dates and time

Engulfy committed:
Add new classes for code to be OOP

Engulfy committed:
Organise into packages

Engulfy committed:
Merge remote-tracking branch 'origin/add-gradle-support'

"Add Gradle Support"

Judy1x4 commented on src/main/java/Engulfy/UI/UI.java:
Shouldn't  the class name be Ui instead of UI?

Engulfy replied:
oh yes thanks for pointing out! will update it later

Judy1x4 commented on src/main/java/Engulfy/Command/MarkCommand.java:
```diff
@@ -0,0 +1,29 @@
+import Engulfy.Storage.Storage;
+import Engulfy.UI.UI;
+
+public class MarkCommand implements Command {
```
Shouldn't there be descriptive header comments? I also noticed they were not present for all public classes/methods in the public in the codebase.

Judy1x4 commented on src/main/java/Engulfy/Errors/EngulfyErrors.java:
Perhaps the class name should be in a singular form as it doesn't represent a collection of objects?

Judy1x4 commented on src/main/java/Engulfy/Storage/Storage.java:
```diff
@@ -0,0 +1,103 @@
+package Engulfy.Storage;
+
+import Engulfy.Errors.EngulfyErrors;
+import Engulfy.Task.*;
```
Shouldn't the imported classes should always be listed explicitly?

Judy1x4 commented on src/main/java/Engulfy/UI/UI.java:
```diff
@@ -0,0 +1,78 @@
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showMarkResult(Task task) {
```
Perhaps showTaskRemoved, showTaskAdded and ShowMarkResult,  ShowUnmarkResult can have a consistent format? e.g. ShowTaskMarked.

Judy1x4 reviewed (commented):
I like how you made use of interfaces and abstract classes for abstraction.

Engulfy committed:
Update file names, integrate gradle in project

Engulfy committed:
Include JUnit Tests for Mark and Unmark Command

Engulfy committed:
Create a JAR file

cerulyean commented on src/main/java/Engulfy/Storage/Storage.java:
```diff
@@ -0,0 +1,107 @@
+                }
+                return task;
+            } else if (type == 'D') {
+                int byIndex = line.lastIndexOf("(by: ");
```
I feel byIndex is too general a description

cerulyean commented on src/main/java/Engulfy/Ui/Ui.java:
```diff
@@ -0,0 +1,80 @@
+    public void showTaskList(List<Task> tasks) {
+        System.out.println("Here are your saved tasks:");
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println("    " + (i + 1) + ". " + tasks.get(i));
+        }
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showNoTasks() {
+        System.out.println("No saved tasks yet!");
+        System.out.println("____________________________________________________________");
+    }
+
+    public String readCommand() {
+        return scanner.nextLine();
+    }
+
+    public void showError(String message) {
+        System.out.println("OOPS! " + message);
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showLoadingError() {
+        System.out.println("Error loading tasks from file. Starting with empty list.");
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showTaskAdded(Task task, int totalTasks) {
+        System.out.println("Got it. I've added this task:");
+        System.out.println("    " + task);
+        System.out.printf("Now you have %d %s in the list.%n", totalTasks, totalTasks == 1 ? "task" : "tasks");
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showTaskRemoved(Task task, int totalTasks) {
+        System.out.println("Noted. I've removed this task:");
+        System.out.println("    " + task);
+        System.out.printf("Now you have %d %s in the list.%n", totalTasks, totalTasks == 1 ? "task" : "tasks");
+        System.out.println("____________________________________________________________");
+    }
```
I like the naming of these methods, they follow the same formatting and describe their function well

cerulyean commented on src/main/java/Engulfy/Ui/Ui.java:
```diff
@@ -0,0 +1,80 @@
+        System.out.println("____________________________________________________________");
+    }
+
+    public void showTaskMark(Task task) {
+        String taskString = task.toString().trim();
+        System.out.print("Nice! I've marked this task as done:\n" +
+                "    " + taskString + "\n" +
+                "____________________________________________________________\n");
+    }
+
+    public void showTaskUnmark(Task task) {
+        String taskString = task.toString().trim();
+        System.out.print("OK, I've marked this task as not done yet:\n" +
+                "    " + taskString + "\n" +
+                "____________________________________________________________\n");
+    }
```
Maybe these methods could be named showTaskMarked() and showTaskUnmarked() to match the style of the earlier methods? I feel showTaskMark could be interpreted as marks in terms of score instead of the action

cerulyean commented on src/main/java/Engulfy/Task/Todo.java:
```diff
@@ -0,0 +1,12 @@
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
```
I like how you use the toString() and size(), to have the same style and function as other classes. Very clear what they do

cerulyean reviewed (commented):
Overall the naming quality of your variables and methods seem clear and descriptive

Engulfy committed:
Create Java Docs for all files

Engulfy committed:
Edit code to follow coding standards

Engulfy committed:
Fix small errors

Engulfy committed:
Resolve conflict while merging

Engulfy committed:
Allow Users to find for task

Engulfy committed:
Resolve conflict while merging

Engulfy committed:
Fix minor errors

Engulfy committed:
Make minor edits

Engulfy committed:
Refine checkstyle errors using CheckStyle

Engulfy committed:
Add GUI to Engulfy bot

Engulfy committed:
Add assertions to document important assumptions

Use Java's assert feature to enforce key assumptions in the code

Assertions have been added to various parts of the program to ensure that
critical conditions hold at runtime. This helps with debugging and
maintaining code correctness by catching unexpected states early.

Key changes:
- Added assertions in TaskList to verify that tasks are not null before
  performing operations.
- Ensured that Storage methods assert file existence before attempting
  to read or write data.
- Updated Parser to assert that input commands are not empty before
  processing them.

These assertions serve as internal documentation and defensive checks,
helping prevent silent failures.

Engulfy committed:
Refactor code to improve quality and maintainability

Improve code quality by refactoring and adhering to best practices

Critically examined the code and made necessary refactoring to enhance
readability, maintainability, and adherence to coding standards.

Key improvements:
- Extracted long methods into smaller, well-defined helper methods to
  improve readability.
- Replaced magic numbers with named constants for better clarity.

These changes enhance code maintainability, making it easier to extend
and debug in the future.

Engulfy committed:
Merge pull request #2 from Engulfy/branch-A-Assertions

Add assertions to document important assumptions

Engulfy committed:
Merge master into branch-A-CodeQuality to update with latest changes

Engulfy committed:
Merge pull request #3 from Engulfy/branch-A-CodeQuality

Refactor code to improve quality and maintainability

Engulfy committed:
Add Tagging Function

Engulfy committed:
Fix minor errors

Engulfy committed:
Improve GUI of bot

Engulfy committed:
Add User Guide for easy reference

Engulfy committed:
Add more personalities to bot

Engulfy committed:
Add an untag function to complement tagging

Engulfy committed:
Update README to contain latest features

Engulfy committed:
Improve tag and untag function

Engulfy committed:
Update README.md

Engulfy committed:
Update README.md

Engulfy committed:
Update README.md

Engulfy committed:
Update README.md
